zerver/tests/test_email_change.py

```diff
@@ -45,15 +50,12 @@ def test_confirm_email_change_when_time_exceeded(self) -> None:
             user_profile=user_profile,
             realm=user_profile.realm,
         )
-        key = generate_key()
+
         date_sent = now() - datetime.timedelta(days=2)
-        Confirmation.objects.create(
-            content_object=obj,
-            date_sent=date_sent,
-            confirmation_key=key,
-            type=Confirmation.EMAIL_CHANGE,
-        )
-        url = confirmation_url(key, user_profile.realm, Confirmation.EMAIL_CHANGE)
+        url = create_confirmation_link(obj, Confirmation.EMAIL_CHANGE)
+        registration_key = url.split("/")[-1]
+        Confirmation.objects.filter(confirmation_key=registration_key).update(date_sent=date_sent)
```

Comment:
This line is hacky; you should either comment that here, or better, avoid doing so.  Probably cleanest would be to replace this whole test with a more end-to-end approach, that does the HTTP request to change the user's email address inside a `mock.patch` for `timezone.now()` to set the time as 2 days ago during that request, finds the email and then confirmation link via `outbox`, and goes from there.

Reply:
End-to-end approach sounds much better.